### ISSUE TITLE
composer update "php:8.2" failing

### ISSUE THREAD

**Issue by cviebrock** [composer/composer#12104]
Following the [docs](https://getcomposer.org/doc/03-cli.md#prohibits-why-not), I'm trying to ensure that my code base will work when upgrading PHP from 8.1 to 8.2.

My `composer.json` is nothing special: there is no platform requirement in the base `require` section, nor is there a `config.platform.php` entry.

As suggested, I run `composer prohibits -r php 8.2` and get the following:

```
Package "php" could not be found with constraint "8.2", results below will most likely be incomplete.                                                      There is no installed package depending on "php" in versions not matching 8.2
Not finding what you were looking for? Try calling `composer update "php:8.2" --dry-run` to get another view on the problem.
```

Looks good so far.  I try running the second command -- `composer update "php:8.2" --dry-run` -- and get the following:

```
Package "php" listed for update is not locked.
Updating dependencies

In RuleSetGenerator.php line 260:
                                                          
  [LogicException]                                        
  Fixed package php 8.2.22 was not added to solver pool.  

Exception trace:
  at phar:///Users/cmv/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:260
 Composer\DependencyResolver\RuleSetGenerator->addRulesForRequest() at phar:///Users/cmv/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:312
 Composer\DependencyResolver\RuleSetGenerator->getRulesFor() at phar:///Users/cmv/bin/composer/src/Composer/DependencyResolver/Solver.php:183
 Composer\DependencyResolver\Solver->solve() at phar:///Users/cmv/bin/composer/src/Composer/Installer.php:508
 Composer\Installer->doUpdate() at phar:///Users/cmv/bin/composer/src/Composer/Installer.php:298
 Composer\Installer->run() at phar:///Users/cmv/bin/composer/src/Composer/Command/UpdateCommand.php:251
 Composer\Command\UpdateCommand->execute() at phar:///Users/cmv/bin/composer/vendor/symfony/console/Command/Command.php:298
 Symfony\Component\Console\Command\Command->run() at phar:///Users/cmv/bin/composer/vendor/symfony/console/Application.php:1040
 Symfony\Component\Console\Application->doRunCommand() at phar:///Users/cmv/bin/composer/vendor/symfony/console/Application.php:301
 Symfony\Component\Console\Application->doRun() at phar:///Users/cmv/bin/composer/src/Composer/Console/Application.php:390
 Composer\Console\Application->doRun() at phar:///Users/cmv/bin/composer/vendor/symfony/console/Application.php:171
 Symfony\Component\Console\Application->run() at phar:///Users/cmv/bin/composer/src/Composer/Console/Application.php:148
 Composer\Console\Application->run() at phar:///Users/cmv/bin/composer/bin/composer:93
 require() at /Users/cmv/bin/composer:29

update [--with WITH] [--prefer-source] [--prefer-dist] [--prefer-install PREFER-INSTALL] [--dry-run] [--dev] [--no-dev] [--lock] [--no-install] [--no-audit] [--audit-format AUDIT-FORMAT] [--no-autoloader] [--no-suggest] [--no-progress] [-w|--with-dependencies] [-W|--with-all-dependencies] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--apcu-autoloader-prefix APCU-AUTOLOADER-PREFIX] [--ignore-platform-req IGNORE-PLATFORM-REQ] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [-m|--minimal-changes] [-i|--interactive] [--root-reqs] [--] [<packages>...]
```

Are the docs incorrect, or is there something I'm missing?

FWIW, output of `composer diagnose`:

```
Checking composer.json: OK
Checking platform settings: OK
Checking git settings: OK git version 2.39.3
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK does not expire
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking Composer version: OK
Checking Composer and its dependencies for vulnerabilities: OK
Composer version: 2.7.9
PHP version: 8.2.22
PHP binary path: /opt/homebrew/Cellar/php@8.2/8.2.22/bin/php
OpenSSL version: OpenSSL 3.3.1 4 Jun 2024
curl version: 8.9.1 libz 1.2.12 ssl OpenSSL/3.3.1 (SecureTransport)
zip: extension present, unzip present, 7-Zip not available
```

**Comment by fredden** [comment]
Please can you provide a copy of your `composer.json` file. It's possible that a package that you depend on has that requirement.

**Comment by cviebrock** [comment]
I'll have to check with my employer, as this is a private project using private packages.  I can guarantee, however, that none of the private packages are requiring `"php":"8.2"` yet.

In the meantime, here is the output of `composer show -D` and `composer show`, with the private packages filtered out.  Maybe that helps?

```
❯ composer show -D
composer/semver                            3.4.2  Semver library that offers utilities, version constraint parsing and validation.
dompdf/dompdf                              2.0.8  DOMPDF is a CSS 2.1 compliant HTML to PDF converter
friendsofphp/php-cs-fixer                  3.64.0 A tool to automatically fix PHP code style
phpstan/phpstan                            1.12.3 PHPStan - PHP Static Analysis Tool
squizlabs/php_codesniffer                  3.10.2 PHP_CodeSniffer tokenizes PHP, JavaScript and CSS files and detects violations of a defined...
```

```
❯ composer show
aws/aws-crt-php                            1.2.6   AWS Common Runtime for PHP
aws/aws-sdk-php                            3.321.8 AWS SDK for PHP - Use Amazon Web Services in your PHP project
clue/ndjson-react                          1.3.0   Streaming newline-delimited JSON (NDJSON) parser and encoder for ReactPHP.
clue/stream-filter                         1.7.0   A simple and modern approach to stream filtering in PHP
codescale/ffmpeg-php                       3.2.2   PHP wrapper for FFmpeg application
composer/pcre                              3.3.1   PCRE wrapping library that offers type-safe preg_* replacements.
composer/semver                            3.4.2   Semver library that offers utilities, version constraint parsing and validation.
composer/xdebug-handler                    3.0.5   Restarts a process without Xdebug.
doctrine/lexer                             3.0.1   PHP Doctrine Lexer parser library that can be used in Top-Down, Recursive Descent Parsers.
dompdf/dompdf                              2.0.8   DOMPDF is a CSS 2.1 compliant HTML to PDF converter
egulias/email-validator                    4.0.2   A library for validating emails against several RFCs
evenement/evenement                        3.0.2   Événement is a very simple event dispatching library for PHP
fidry/cpu-core-counter                     1.2.0   Tiny utility to get the number of CPU cores.
friendsofphp/php-cs-fixer                  3.64.0  A tool to automatically fix PHP code style
guzzlehttp/guzzle                          7.9.2   Guzzle is a PHP HTTP client library
guzzlehttp/promises                        2.0.3   Guzzle promises library
guzzlehttp/psr7                            2.7.0   PSR-7 message implementation that also provides common utility methods
http-interop/http-factory-guzzle           1.2.0   An HTTP Factory using Guzzle PSR7
jean85/pretty-package-versions             2.0.6   A library to get pretty versions strings of installed dependencies
league/climate                             3.8.2   PHP's best friend for the terminal. CLImate allows you to easily output colored text, spec...
masterminds/html5                          2.9.0   An HTML5 parser and serializer.
mtdowling/jmespath.php                     2.8.0   Declaratively specify how to extract elements from a JSON document
pear/console_commandline                   1.2.6   A full featured command line options and arguments parser.
pear/console_getopt                        1.4.3   More info available on: http://pear.php.net/package/Console_Getopt
pear/pear-core-minimal                     1.10.15 Minimal set of PEAR core files to be used as composer dependency
pear/pear_exception                        1.0.2   The PEAR Exception base class.
pear/text_password                         1.2.2   More info available on: http://pear.php.net/package/Text_Password
phenx/php-font-lib                         0.5.6   A library to read, parse, export and make subsets of different types of font files.
phenx/php-svg-lib                          0.5.4   A library to read, parse and export to PDF SVG files.
php-http/client-common                     2.7.1   Common HTTP Client implementations and tools for HTTPlug
php-http/discovery                         1.19.4  Finds and installs PSR-7, PSR-17, PSR-18 and HTTPlug implementations
php-http/httplug                           2.4.0   HTTPlug, the HTTP client abstraction for PHP
php-http/message                           1.16.1  HTTP Message related tools
php-http/message-factory                   1.1.0   Factory interfaces for PSR-7 HTTP Message
php-http/promise                           1.3.1   Promise used for asynchronous HTTP requests
phpstan/phpstan                            1.12.3  PHPStan - PHP Static Analysis Tool
psr/container                              2.0.2   Common Container Interface (PHP FIG PSR-11)
psr/event-dispatcher                       1.0.0   Standard interfaces for event handling.
psr/http-client                            1.0.3   Common interface for HTTP clients
psr/http-factory                           1.1.0   PSR-17: Common interfaces for PSR-7 HTTP message factories
psr/http-message                           2.0     Common interface for HTTP messages
psr/log                                    3.0.1   Common interface for logging libraries
ralouphie/getallheaders                    3.0.3   A polyfill for getallheaders.
react/cache                                1.2.0   Async, Promise-based cache interface for ReactPHP
react/child-process                        0.6.5   Event-driven library for executing child processes with ReactPHP.
react/dns                                  1.13.0  Async DNS resolver for ReactPHP
react/event-loop                           1.5.0   ReactPHP's core reactor event loop that libraries can use for evented I/O.
react/promise                              3.2.0   A lightweight implementation of CommonJS Promises/A for PHP
react/socket                               1.16.0  Async, streaming plaintext TCP/IP and secure TLS socket server and client connections for ...
react/stream                               1.4.0   Event-driven readable and writable streams for non-blocking I/O in ReactPHP
sabberworm/php-css-parser                  8.6.0   Parser for CSS Files written in PHP
sebastian/diff                             5.1.1   Diff implementation
seld/cli-prompt                            1.0.4   Allows you to prompt for user input on the command line, and optionally hide the character...
sentry/sdk                                 3.6.0   This is a metapackage shipping sentry/sentry with a recommended HTTP client.
sentry/sentry                              3.22.1  A PHP SDK for Sentry (http://sentry.io)
squizlabs/php_codesniffer                  3.10.2  PHP_CodeSniffer tokenizes PHP, JavaScript and CSS files and detects violations of a define...
symfony/console                            6.4.11  Eases the creation of beautiful and testable command line interfaces
symfony/deprecation-contracts              3.5.0   A generic function and convention to trigger deprecation notices
symfony/event-dispatcher                   6.4.8   Provides tools that allow your application components to communicate with each other by di...
symfony/event-dispatcher-contracts         3.5.0   Generic abstractions related to dispatching event
symfony/filesystem                         6.4.9   Provides basic utilities for the filesystem
symfony/finder                             6.4.11  Finds files and directories via an intuitive fluent interface
symfony/http-client                        6.4.11  Provides powerful methods to fetch HTTP resources synchronously or asynchronously
symfony/http-client-contracts              3.5.0   Generic abstractions related to HTTP clients
symfony/mailer                             5.4.41  Helps sending emails
symfony/mime                               6.4.11  Allows manipulating MIME messages
symfony/options-resolver                   6.4.8   Provides an improved replacement for the array_replace PHP function
symfony/polyfill-ctype                     1.31.0  Symfony polyfill for ctype functions
symfony/polyfill-intl-grapheme             1.31.0  Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-idn                  1.31.0  Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer           1.31.0  Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring                  1.31.0  Symfony polyfill for the Mbstring extension
symfony/polyfill-php80                     1.31.0  Symfony polyfill backporting some PHP 8.0+ features to lower PHP versions
symfony/polyfill-php81                     1.31.0  Symfony polyfill backporting some PHP 8.1+ features to lower PHP versions
symfony/process                            6.4.8   Executes commands in sub-processes
symfony/service-contracts                  3.5.0   Generic abstractions related to writing services
symfony/stopwatch                          6.4.8   Provides a way to profile code
symfony/string                             6.4.11  Provides an object-oriented API to strings and deals with bytes, UTF-8 code points and gra...
symfony/yaml                               5.4.43  Loads and dumps YAML files
```

**Comment by Seldaek** [comment]
I think the correct way to check for compatibility with 8.2 would be to set the `composer config platform.php 8.2.23` then run a `composer update --lock --dry-run` to see if your current lock file is compatible, or `composer update --dry-run` to see if you could find a compatible set of dependencies.

**Comment by stof** [comment]
`composer update "php:8.2"` does not make sense, as `php` is a platform package that cannot be updated by composer itself.

**Comment by stof** [comment]
@Seldaek maybe composer should detect the case of passing platform packages to the `update` command, to provide a better error message than the failure caused by the fixed package not being added in the pool.

**Comment by cviebrock** [comment]
@Seldaek and @stof thanks for the comments and suggestions.

I agree that composer can't actually update the platform ... nor would I expect it to!  I think the main reason for posting this issue was to let you know that the action suggested by composer throws an error.  However, if that suggestion doesn't make sense, maybe composer shouldn't suggest it in the first place?

Thanks!

**Comment by Seldaek** [comment]
Yeah we should handle this better both in update and also not suggest this nonsense in prohibits I guess.. I'll see how easy that'd be to detect.

**Comment by stof** [comment]
@Seldaek while you are at it, it may make sense to remove the warning saying that the package cannot be found in repositories in `prohibits` when the package is a platform package.

**Comment by Seldaek** [comment]
Yeah I've fixed that and more, see #12110

**Comment by ronssij** [comment]
[
![IMG_3401](https://github.com/user-attachments/assets/857ab59c-11f9-4a67-8cb2-654ad14dab65)
](url)
![image](https://github.com/user-attachments/assets/8065406b-1eeb-4fc2-b05b-4f5a2703d1ae)

I have experienced this 3 days ago. Not sure if this still exists today but will try the newest version 2.8.1 on Monday. What we did for now was to force rolling back composer's version to 7.29.

**Comment by xabbuh** [comment]
@ronssij This looks like a different issue and should be fixed with #12152
